[Connor] iP

connlim opened the pull request:
# Piggy 
> “Object-oriented programming is an exceptionally bad idea which could only have originated in California.” – Edsger Dijkstra ([source](https://harmful.cat-v.org/software/OO_programming/))

## Types of Tasks
- Todo
- Deadline (supports a by date)
- Event (supports an at date)

## Features
1. Add tasks
2. Delete tasks
3. Mark tasks as done/undone
4. All tasks are loaded from and saved to disk automatically 💾
5. **FAST** to use
6. ~~Simple~~Usable interface e.g. `todo My Todo task`

## Planned Features
- [x] Find tasks
- [ ] Proper GUI
- [ ] Automatic style checking
- [ ] Varargs

## Data File Loading
```java
public static void main(String[] args) {
    new Piggy("./tasks.txt").run();
}
```

Commit:
Add Gradle support

connlim committed:
Add skeletal chatbot

connlim committed:
Add ability to store and display text

connlim committed:
Add ability to mark as done

connlim committed:
Add toString function for Task

connlim committed:
Add task types

connlim committed:
Add ability to add different tasks

connlim committed:
Update gitignore to ignore actual text file

connlim committed:
Add semi-automated testing

connlim committed:
Add error handling

connlim committed:
Add ability to delete task

connlim committed:
Add tasks list to gitignore

connlim committed:
Add ability to read and write tasks from file

connlim committed:
Add datetime parsing capability

connlim committed:
Add datetime format info to intro

connlim committed:
Merge branch 'branch-Level-7'

connlim committed:
Merge branch 'branch-Level-8'

connlim committed:
Update TaskWriter to use TaskWithDate

connlim committed:
Rename TaskWriter to Storage

connlim committed:
Fix datetime parsing from file

connlim committed:
Move all user interactions to Ui class

connlim committed:
Extract task list to TaskList class

connlim committed:
Extract command parsing to Parser class

connlim committed:
Extract main function to Duke class

connlim committed:
Rename Duke to Piggy

connlim committed:
Divide classes into packages

connlim committed:
Merge branch 'add-gradle-support'

connlim committed:
Update gradle config for Piggy

connlim committed:
Add JUnit tests

connlim committed:
Add JavaDoc comments

connlim committed:
Update code to follow coding standard

connlim committed:
Add ability to find tasks

connlim committed:
Merge branch 'branch-A-JavaDoc'

connlim committed:
Merge branch 'branch-A-CodingStandard'

connlim committed:
Merge branch 'branch-Level-9'

mohamedsaf1 commented on src/main/java/piggy/Piggy.java:
```diff
@@ -0,0 +1,44 @@
+            isExit = Parser.parse(command, ui, taskList);
+            storage.writeTasks(taskList);
+        }
+        ui.showBye();
```
good initiative greeting "bye"

mohamedsaf1 commented on src/main/java/piggy/Storage.java:
```diff
@@ -0,0 +1,126 @@
+                String[] data = sc.nextLine().split(",");
+
+                Task task;
+                boolean done = Boolean.parseBoolean(data[1]);
```
does not follow coding standard, better naming would be for boolean "isDone"

connlim replied:
Agreed, will be fixed.

mohamedsaf1 commented on src/main/java/piggy/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     * @param task             The removed task to show.
+     * @param noRemainingTasks The number of remaining tasks.
+     */
+    void showTaskRemoved(Task task, int noRemainingTasks) {
```
variable "noRemainingTasks" sounds like there are no remaining tasks rather than number of remaining tasks.

connlim replied:
Agreed

peppapighs commented on src/main/java/piggy/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     * @param task             The removed task to show.
+     * @param noRemainingTasks The number of remaining tasks.
+     */
+    void showTaskRemoved(Task task, int noRemainingTasks) {
```
Consider renaming this method to `showRemoveTask` or `showRemovedTask`? The ordering of the words makes the method name seems unclear to me.

connlim replied:
Yeah that makes more sense thanks

peppapighs commented on src/main/java/piggy/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     * @param task             The added task to show.
+     * @param noRemainingTasks The number of remaining tasks.
+     */
+    void showTaskAdded(Task task, int noRemainingTasks) {
```
Consider renaming this method to `showAddTask` or `showAddedTask`? The ordering of the words makes the method name seems unclear to me.

peppapighs commented on src/main/java/piggy/Parser.java:
```diff
@@ -0,0 +1,74 @@
+        } else {
+            Task task;
+            try {
+                if (command.startsWith("todo")) {
+                    if (command.length() < 6) {
+                        throw new DukeException("The description of a todo cannot be empty.");
+                    }
+                    task = new Todo(command.substring(5));
+                } else if (command.startsWith("deadline")) {
+                    if (command.length() < 10) {
+                        throw new DukeException("The description of a deadline cannot be empty.");
+                    } else if (!command.contains("/by")) {
+                        throw new DukeException("A deadline must contain a /by");
+                    }
+                    String[] split = command.substring(9).split(" /by ");
+                    task = new Deadline(split[0], split[1]);
+                } else if (command.startsWith("event")) {
+                    if (command.length() < 7) {
+                        throw new DukeException("The description of an event cannot be empty.");
+                    } else if (!command.contains("/at")) {
+                        throw new DukeException("An event must contain an /at");
+                    }
+                    String[] split = command.substring(6).split(" /at ");
+                    task = new Event(split[0], split[1]);
+                } else {
+                    throw new DukeException("I'm sorry, but I don't know what that means :-(");
+                }
+                taskList.add(task);
+                ui.showTaskAdded(task, taskList.size());
```
Great uses of return early pattern to reduce nested statements.

peppapighs reviewed (commented):
Overall, the code looks good to me. Just some minor method naming that I feel are a bit unclear.

connlim committed:
Add assertions

connlim committed:
Add stream

connlim committed:
Merge pull request #2 from connlim/branch-A-Assertions

Add assertions

connlim committed:
Merge pull request #1 from connlim/branch-A-Streams

Add streams

connlim committed:
Update Duke to Piggy in readme

The readme references Duke in several places. This is outdated.
The readme has been fixed by changing all references to Duke to
the name of the new app, Piggy. This will make the readme easier
to read.

connlim committed:
Update the welcome message in the readme

The current welcome message is a simple ASCII art
for duke.

This is outdated as the app is now called Piggy, and
the welcome message has been changes.

Let's change the code block containing the Duke ASCII
art to contain the new Piggy welcome message.

This will prevent user confusion.

connlim committed:
Add javafx

connlim committed:
Merge branch 'branch-Level-10'

connlim committed:
Format code to comply with codestyle

connlim committed:
Merge branch 'branch-A-CodeQuality'

connlim committed:
Add ability to see schedule for a date

connlim committed:
Merge branch 'branch-B-ViewSchedules'

connlim committed:
Add window title

connlim committed:
Add UI image

connlim committed:
Add user guide

connlim committed:
Merge branch 'branch-A-UserGuide'
